### PR TITLE
kv: Rocksdb stats

### PR DESCRIPTION
**Pull request by varadakari**
This is PR is an extension to #11729 

Added extended stats from Rocksdb and collecting the stats only when rocksdb_perf is enabled.
rocksdb_perf should be enabled while initializing the store to collect all the stats.

### CODE REVIEW

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -363,6 +366,26 @@ void RocksDBStore::close()
...
+      if (p1 != p2) {
+	std::string substr = stats.substr(p2, p1 - p2 + 1);
+	substr = "  " + substr;
+	f->write_raw_data(substr.data());
+      } else
+	f->write_raw_data("\n");
+      p2 = ++p1;
+    }
+    f->write_raw_data("\n");
+    f->close_section();
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: I don't think we want to be doing nay write_raw_data or else we're generate invalid json and xml both.  If the stats are preformatted then we can do a single dump_string().

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -1935,6 +1935,9 @@ bool OSD::asok_command(string command, cmdmap_t& cmdmap, string format,
     f->dump_bool("success", success);
     f->dump_int("value", value);
     f->close_section();
+  } else if (command == "bluestore_db_statistics") {
+    if (store->get_type() == "bluestore")
+      store->get_db_statistics(f);
```
Comment (liewegas) [inline src/osd/OSD.cc]: dump_objecststore_stats ?

and let's just make get_db_statistics a no-op in ObjectStore and overlaod only in bluestore. Then we don't need the type conditional here.

File: src/kv/RocksDBStore.cc [hunk 3]
```diff
@@ -371,6 +371,7 @@ void RocksDBStore::get_statistics(Formatter *f) {
   bool status = db->GetProperty("rocksdb.stats", &stats);
   if (status) {
     f->open_object_section("bluestore rocksdb statistics");
+    f->write_raw_data("\n We usually observe an overhead of 5%-10% for this command. Be carefully before using\n\n");
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: no write_raw_data, please!

File: src/common/config_opts.h [hunk 4]
```diff
@@ -1071,6 +1071,7 @@ OPTION(bluestore_debug_inject_read_err, OPT_BOOL, false)
 OPTION(bluestore_debug_randomize_serial_transaction, OPT_INT, 0)
 OPTION(bluestore_inject_wal_apply_delay, OPT_FLOAT, 0)
 OPTION(bluestore_shard_finishers, OPT_BOOL, false)
+OPTION(bluestore_enable_db_collect_statistics, OPT_BOOL, false) //For rocksdb, this behavior will be an overhead of 5%~10%.
```
Comment (liewegas) [inline src/common/config_opts.h]: bluestore_collect_kv_stats

File: src/common/config_opts.h [hunk 5]
```diff
@@ -1072,6 +1072,8 @@ OPTION(bluestore_debug_randomize_serial_transaction, OPT_INT, 0)
 OPTION(bluestore_inject_wal_apply_delay, OPT_FLOAT, 0)
 OPTION(bluestore_shard_finishers, OPT_BOOL, false)
 OPTION(bluestore_enable_db_collect_statistics, OPT_BOOL, false) //For rocksdb, this behavior will be an overhead of 5%~10%.
+OPTION(bluestore_enable_db_collect_extended_statistics, OPT_BOOL, false) //For rocksdb, this behavior will be an overhead of 5%~10%.
+OPTION(bluestore_enable_db_collect_memory_statistics, OPT_BOOL, false) //For rocksdb, this behavior will be an overhead of 5%~10%.
```
Comment (liewegas) [inline src/common/config_opts.h]: bluestore_collect_kv_stats_*

File: src/kv/RocksDBStore.cc [hunk 6]
```diff
@@ -367,11 +370,15 @@ void RocksDBStore::close()
 }
 
 void RocksDBStore::get_statistics(Formatter *f) {
+  if (!g_conf->rocksdb_perf)  {
+    f->write_raw_data("\n RocksDB perf is disabled, can't probe for stats\n\n");
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: no write_raw_data.. just leave it blank?

File: src/kv/RocksDBStore.cc [hunk 7]
```diff
@@ -385,6 +392,44 @@ void RocksDBStore::get_statistics(Formatter *f) {
     f->write_raw_data("\n");
     f->close_section();
   }
+  if (g_conf->bluestore_enable_db_collect_extended_statistics) {
+    if (dbstats) {
+      f->open_object_section("bluestore rocksdb extended statistics");
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: s/ /_/g

File: src/kv/RocksDBStore.cc [hunk 8]
```diff
@@ -385,6 +392,44 @@ void RocksDBStore::get_statistics(Formatter *f) {
...
+      f->write_raw_data("\n");
+      f->write_raw_data(dbstats->ToString().c_str());
+      f->write_raw_data("\n");
```
Comment (liewegas) [inline src/kv/RocksDBStore.cc]: no write_raw_data...

**Comment by varadakari** [comment]
@liewegas compaction stats are printed in tabular format, if we use dump_string() it will be tough to make out the output. So had to use write_raw_data().
Here is the sample format what is dumped in the stats:

{
** Compaction Stats [default] **
Level    Files   Size(MB} Score Read(GB}  Rn(GB} Rnp1(GB} Write(GB} Wnew(GB} Moved(GB} W-Amp Rd(MB/s} Wr(MB/s} Comp(sec} Comp(cnt} Avg(sec} KeyIn KeyDrop
----------------------------------------------------------------------------------------------------------------------------------------------------------
  L1      1/0       0.00   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0         0         0    0.000       0      0
 Sum      1/0       0.00   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0         0         0    0.000       0      0
 Int      0/0       0.00   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0         0         0    0.000       0      0
Uptime(secs): 63.8 total, 63.8 interval
Flush(GB): cumulative 0.000, interval 0.000
AddFile(GB): cumulative 0.000, interval 0.000
AddFile(Total Files): cumulative 0, interval 0
AddFile(L0 Files): cumulative 0, interval 0
AddFile(Keys): cumulative 0, interval 0

**Comment by varadakari** [comment]
@liewegas  will squash and cleanup the commits if you are okay to use write_raw_data() else will switch to dump_string() and have a helper script to format the stats.

**Comment by markhpc** [comment]
varada:

fwiw, I have a similar helper script here for some of the other statistics:

https://github.com/ceph/cbt/blob/master/tools/ceph_rocksdb_log_parser.py

**Comment by varadakari** [comment]
@markhpc script looks good mark.  we can use that for parsing the output from stats. But if we want to read it clearly(rocksdb prints in the above mentioned format and using dump_string() distorts that), then write_raw_data() is the way(i don't know if there any other API can format it clean). 

I will change to dump_string() if we feel, we need to have in the same format as rocksdb, then we can switch back.

**Comment by dachary** [comment]
jenkins test this please (https://jenkins.ceph.com/job/ceph-pull-requests/14954/console has cephtool-test-mds.sh and unittest_ceph_crypto failing) not sure how they can be related to this PR

**Comment by liewegas** [comment]
see https://github.com/ceph/ceph/pull/12287

**Pull request closed without merging**